File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    static List<List<Integer>> answer = new ArrayList<>();
    private static void  helper(int[] nums,int target, int currentsum , List<Integer> ds,int index){
        if(currentsum == target){
            answer.add(new ArrayList<>(ds));
            return;
        }
        if(currentsum > target || index >= nums.length){
            return;
        }
        ds.add(nums[index]);
        helper(nums,target,currentsum+nums[index],ds,index);
        ds.remove(ds.size()-1);
        helper(nums,target,currentsum,ds,index+1);

    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        answer.clear();
        helper(candidates,target,0,new ArrayList<>(),0);
        return answer;
    }
}

File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<vector<int>> result;
    void helper(int i,int currentsum,int target,vector<int> ds,vector<int> nums){
        if(i>=nums.size()) return;
        if(currentsum > target) return;
        if(target == currentsum){
            result.push_back(ds);
            return;
        }
        ds.push_back(nums[i]);
        helper(i,currentsum+nums[i],target,ds,nums);
        ds.pop_back();
        helper(i+1,currentsum,target,ds,nums);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> ds;
        helper(0,0,target,ds,candidates);
        return result;
    }
};